diff --git a/fs/dcache.c b/fs/dcache.c
index e88cf0554..6c35d60e4 100644
--- a/fs/dcache.c
+++ b/fs/dcache.c
@@ -2890,9 +2890,10 @@ static void __d_move(struct dentry *dentry, struct dentry *target,
  */
 void d_move(struct dentry *dentry, struct dentry *target)
 {
-	write_seqlock(&rename_lock);
+	extern int num_policy;
+	bpf_write_seqlock(&rename_lock, num_policy);
 	__d_move(dentry, target, false);
-	write_sequnlock(&rename_lock);
+	bpf_write_sequnlock(&rename_lock, num_policy);
 }
 EXPORT_SYMBOL(d_move);
 
diff --git a/kernel/bpf/inode.c b/kernel/bpf/inode.c
index a70f7209c..95a964262 100644
--- a/kernel/bpf/inode.c
+++ b/kernel/bpf/inode.c
@@ -701,3 +701,79 @@ static int __init bpf_init(void)
 	return ret;
 }
 fs_initcall(bpf_init);
+
+
+#include "kpatch-macros.h"
+#define MAX_POLICY 5
+
+static inline __u64 ptr_to_u64(const void *ptr)
+{
+    return (__u64) (unsigned long) ptr;
+}
+
+static void *get_pinned_bpf_obj(const char *pathname){
+	struct inode *inode;
+	struct path path;
+	void *raw;
+	int ret;
+
+	// Let's get BPF prog 1
+	ret = kern_path(pathname, LOOKUP_FOLLOW, &path);
+	if (ret){
+		printk("[syncord] %s failed\n", pathname);
+		return ERR_PTR(ret);
+	}
+
+	inode = d_backing_inode(path.dentry);
+	ret = inode_permission(inode, ACC_MODE(2));
+	if(ret){
+		printk("[syncord] perm error\n");
+		path_put(&path);
+		return ERR_PTR(ret);
+	}
+
+	raw = bpf_any_get(inode->i_private, BPF_TYPE_PROG);
+	if(!IS_ERR(raw)){
+		touch_atime(&path);
+	}
+	else{
+		printk("[syncord] raw error\n");
+		path_put(&path);
+		return ERR_PTR(ret);
+	}
+
+	path_put(&path);
+	return raw;
+}
+
+static int pre_patch_callback(patch_object *obj)
+{
+	extern int num_policy;
+	extern void *bpf_prog_lock_to_enter_slowpath[MAX_POLICY];
+	extern void *bpf_prog_lock_acquired[MAX_POLICY];
+
+	bpf_prog_lock_to_enter_slowpath[num_policy] = get_pinned_bpf_obj("/sys/fs/bpf/amp_cont_lock");
+	if(IS_ERR(bpf_prog_lock_to_enter_slowpath[num_policy])){
+		printk("[syncord] bpf_policy failed\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+static void post_unpatch_callback(patch_object *obj)
+{
+	extern int num_policy;
+	extern void *bpf_prog_lock_to_enter_slowpath[MAX_POLICY];
+	extern void *bpf_prog_lock_acquired[MAX_POLICY];
+
+	bpf_prog_lock_to_enter_slowpath[num_policy] = NULL;
+	bpf_prog_lock_acquired[num_policy] = NULL;
+
+	num_policy--;
+
+	klp_shadow_free_all(0, NULL);
+}
+
+KPATCH_PRE_PATCH_CALLBACK(pre_patch_callback);
+KPATCH_POST_UNPATCH_CALLBACK(post_unpatch_callback);
diff --git a/kernel/locking/qspinlock.c b/kernel/locking/qspinlock.c
index cfdd020be..10b30e33d 100644
--- a/kernel/locking/qspinlock.c
+++ b/kernel/locking/qspinlock.c
@@ -29,6 +29,9 @@
  * Include queued spinlock statistics code
  */
 #include "qspinlock_stat.h"
+#include <linux/lock_policy.h>
+#include <linux/filter.h>
+#include <linux/livepatch.h>
 
 /*
  * The basic principle of a queue-based spinlock can best be understood
@@ -225,6 +228,14 @@ static void syncord_lock_released(struct qspinlock *lock, int policy_id)
 // Fastpath APIs
 static void syncord_to_enter_slowpath(struct qspinlock *lock, struct mcs_spinlock *node, int policy_id)
 {
+	struct bpf_prog *prog;
+	prog = bpf_prog_lock_to_enter_slowpath[policy_id];
+
+	struct lock_policy_args args;
+	args.lock_ptr = (unsigned long)lock;
+
+	int ret = BPF_PROG_RUN(prog, &args);
+
 	return;
 }
 
